### ISSUE TITLE
fp64 shader emulation not working under Apple M1 Max

### ISSUE THREAD

**Issue by Swoorup** [visgl/luma.gl#1764]
Using the fp64 modules, it appears as the precision are lost under Apple M1 MAX, as if I were using fp32. The values sent to the shader do appear correct. I am not sure of the best way to reproduce as it works across an older Intel laptop or my linux PC. 

Curious if anyone has encountered the same.

**Comment by ibgreen** [comment]
Yes I have noticed this as well, but haven't had time to look into it.

The problem is generally that the GPU drivers (shader compilers) optimize the shaders so aggressively that they disable the required computational tricks.

I believe that there is a flag for nVidia to add additional workarounds to prevent optimization and I did make a quick try at enabling that for apple chips without success.

Someone may need to do a deeper dive here.

**Comment by Swoorup** [comment]
Interestingly this simple example doesn't work either
https://prideout.net/emulating-double-precision

**Comment by Swoorup** [comment]
This appears to be an ANGLE issue. I got it fixed using a workaround `invariant gl_Position`
mentioned in here: https://bugs.webkit.org/show_bug.cgi?id=237434

**Comment by ibgreen** [comment]
@swoorup That sounds very promising, thanks for researching the issue!

**Comment by donmccurdy** [comment]
> Someone may need to do a deeper dive here.

I can reproduce the issue at https://prideout.net/emulating-double-precision on an Apple M2 Pro with macOS 14 Sonoma, but I'm not sure where to start digging into the issue on visgl. Is that still of interest, and would someone be able to share an example or failing unit test? Thanks!

**Comment by ibgreen** [comment]
We should re-enable the tests in luma.gl and then we should check if @Swoorup 's suggestion in this thread works.

**Comment by donmccurdy** [comment]
Update – With lots of development toward v9, the Transform class was a bit behind and FP64 tests couldn't run without that. I've got a minimal version of Transform running again, and have restored the FP64 tests in https://github.com/visgl/luma.gl/pull/1879. With that we can reproduce the problem easily here.

I'd hoped adding `invariant varying vec2 result;`...

https://github.com/visgl/luma.gl/blob/8d1b5b4d013f7f7a9d4b1ebbc23a5e5efb2dc4e8/modules/shadertools/test/modules-webgl1/fp64/fp64-test-utils-transform.ts#L14-L24

... would fix the problem on Apple GPUs, but unfortunately I don't see any change on my device (Apple M2 Pro). To re-enable any of the failing tests on Apple GPUs and reproduce the failure, just edit the test cases here ...

https://github.com/visgl/luma.gl/blob/8d1b5b4d013f7f7a9d4b1ebbc23a5e5efb2dc4e8/modules/shadertools/test/modules-webgl1/fp64/fp64-arithmetic-transform.spec.ts#L11-L14

... and run `yarn test browser`. I've also tried switching over to some other fp64 arithmetic implementations, found in https://blog.cyclemap.link/2011-06-09-glsl-part2-emu/, but the results were slightly worse than our current state. I've forced the `Choose ANGLE graphics backend` option to "Metal" in `chrome://flags/`, to ensure it doesn't switch back to OpenGL.

**Comment by donmccurdy** [comment]
Updates in #1882. We can fix the issue with the `invariant` specifier on Apple M1/M2 in Chrome, when ANGLE's OpenGL backend is enabled. For unknown reasons, the specifier doesn't seem to help when ANGLE's Metal backend is enabled. The default is currently up to the browser. No effect in Safari, which still fails.

**Comment by donmccurdy** [comment]
I've added a small, reproducible example to the Webkit bug, without a dependency on lumagl:

- Demo: https://fp64-arithmetic.donmccurdy.com/
- Code: https://github.com/donmccurdy/fp64-arithmetic

**Comment by Swoorup** [comment]
I wonder if it is because it is disabled on anything not Mac 11 (Angle Metal)? I haven't got time to test compiling with it turned on though.

https://github.com/google/angle/blob/5678ad09aa3b72c5b7c7b05c49c6c8b3ed443487/src/libANGLE/renderer/metal/mtl_utils.mm#L891-L897

**Comment by ibgreen** [comment]
This is resolved on master / v10. The changes are substantial, unlikely to be back ported.
